library/core/src/num/mod.rs

```diff
@@ -846,37 +870,32 @@ fn from_str_radix<T: FromStrRadixHelper>(src: &str, radix: u32) -> Result<T, Par
     };
 
     let mut result = T::from_u32(0);
-    if is_positive {
-        // The number is positive
-        for &c in digits {
-            let x = match (c as char).to_digit(radix) {
-                Some(x) => x,
-                None => return Err(PIE { kind: InvalidDigit }),
-            };
-            result = match result.checked_mul(radix) {
-                Some(result) => result,
-                None => return Err(PIE { kind: PosOverflow }),
-            };
-            result = match result.checked_add(x) {
-                Some(result) => result,
-                None => return Err(PIE { kind: PosOverflow }),
-            };
+
+    if radix <= 16 && src.len() <= mem::size_of::<T>() * 2 - (is_signed_ty && is_positive) as usize
+    {
+        // SAFETY: Consider the highest radix of 16:
+        // `u8::MAX` is `ff` which is 2 characters, `u16::MAX` out is `ffff` which is 4 characters
```

Comment:
```suggestion
        // `u8::MAX` is `ff` (2 characters), `u16::MAX` is `ffff` (4 characters)
```

Can simplify this with less words and easier to understand.